### PR TITLE
Overlay controlled

### PR DESCRIPTION
**Pull request by lukeac123**
Resolved issue - https://github.com/jpmorganchase/salt-ds/issues/3077
The value passed to useFloatingUI changes depending on if the component is in a controlled state
onOpenChange is needed as a prop when component is in controlled state
onClose removed as a prop

- Added cypress story testing overlay in the controlled state 
- Added QA test

### CODE REVIEW

File: packages/lab/src/overlay/Overlay.tsx [hunk 1]
```diff
@@ -21,15 +15,11 @@ import {
 
 export interface OverlayProps extends ComponentPropsWithoutRef<"div"> {
   open?: boolean;
-  onOpenChange?: (event: SyntheticEvent, newOpen: boolean) => void;
```
Comment (origami-z) [inline packages/lab/src/overlay/Overlay.tsx:24]: Better to keep event whenever possible

File: .changeset/plenty-schools-fix.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+---
+"@salt-ds/lab": minor
+---
+
+Remove `onClose` prop from `Overlay`, onOpenChange prop required in controlled state
```
Comment (origami-z) [inline .changeset/plenty-schools-fix.md:5]: Does this mean uncontrolled can't be used?

File: packages/lab/src/__tests__/__e2e__/overlay/Overlay.cy.tsx [hunk 3]
```diff
@@ -32,13 +32,11 @@ describe("GIVEN an Overlay", () => {
     });
 
     it("THEN it should remain open until outside Overlay click or close button click", () => {
-      const closeSpy = cy.stub().as("closeSpy");
-      cy.mount(<Default onClose={closeSpy} />);
```
Comment (origami-z) [inline packages/lab/src/__tests__/__e2e__/overlay/Overlay.cy.tsx:36]: Test `onOpenChange` instead?

File: packages/lab/src/overlay/Overlay.tsx (hunk 1 above)
Comment (joshwooding) [reply packages/lab/src/overlay/Overlay.tsx:24]: This was done to match other components in core

File: packages/lab/src/overlay/Overlay.tsx [hunk 4]
```diff
@@ -40,14 +30,13 @@ export const Overlay = forwardRef<HTMLDivElement, OverlayProps>(
...
       id: idProp,
-      onClose,
       ...rest
     } = props;
 
     const id = useId(idProp);
     const arrowRef = useRef<SVGSVGElement | null>(null);
 
-    const [openState, setOpenState] = useControlled({
+    const [openState, setOpenState, isControlled] = useControlled({
```
Comment (joshwooding) [inline packages/lab/src/overlay/Overlay.tsx:39]: This seems unnecessary and doesn't match how we've done this elsewhere. We should create a handler in the code that sets the state and calls onOpenChange. Uncontrolled components should still call the event handlers.

File: packages/lab/stories/overlay/overlay.stories.tsx [hunk 5]
```diff
@@ -215,12 +215,12 @@ export const WithActions = () => {
   const [show, setShow] = React.useState(false);
   const id = "overlay-with-actions";
 
+  const onOpenChange = () => setShow(false);
```
Comment (joshwooding) [inline packages/lab/stories/overlay/overlay.stories.tsx:218]: This should be updated to read the newOpen parameter

File: site/src/examples/overlay/WithActions.tsx [hunk 6]
```diff
@@ -108,12 +108,12 @@ export const WithActions = () => {
   const [show, setShow] = React.useState(false);
   const id = useId();
 
+  const onOpenChange = () => setShow(false);
```
Comment (joshwooding) [inline site/src/examples/overlay/WithActions.tsx:111]: Same here

File: .changeset/plenty-schools-fix.md (hunk 2 above)
Comment (lukeac123) [reply .changeset/plenty-schools-fix.md:5]: The overlay manages the open state by default in an uncontrolled state - if the user wants something additional to happen when the close button is clicked then yes the overlay would have to be in a controlled state

File: packages/lab/src/__tests__/__e2e__/overlay/Overlay.cy.tsx [hunk 7]
```diff
@@ -50,8 +50,9 @@ describe("GIVEN an Overlay", () => {
     });
 
     it("THEN it should trap focus within Overlay once opened", () => {
-      cy.mount(<Default open />);
```
Comment (origami-z) [inline packages/lab/src/__tests__/__e2e__/overlay/Overlay.cy.tsx:53]: Is controlled version tested?

File: packages/lab/src/overlay/OverlayContext.ts [hunk 8]
```diff
@@ -20,14 +20,13 @@ type FloatingStyleProps = {
...
   openState: boolean;
-  setOpen: (event: SyntheticEvent, newOpen: boolean) => void;
   floatingStyles: FloatingStyleProps;
   placement: Placement;
   arrowProps: FloatingArrowProps;
   context: FloatingContext;
   reference?: (node: ReferenceType | null) => void;
   floating?: (node: HTMLElement | null) => void;
-  handleCloseButton: (event: SyntheticEvent) => void;
+  handleCloseButton: () => void;
```
Comment (origami-z) [inline packages/lab/src/overlay/OverlayContext.ts:29]: Should be be something related to click?

File: packages/lab/src/__tests__/__e2e__/overlay/Overlay.cy.tsx (hunk 7 above)
Comment (lukeac123) [reply packages/lab/src/__tests__/__e2e__/overlay/Overlay.cy.tsx:53]: Added tests to make sure the onOpenChange function is called correctly in a controlled state

File: packages/lab/src/__tests__/__e2e__/overlay/Overlay.cy.tsx [hunk 9]
```diff
@@ -112,4 +115,26 @@ describe("GIVEN an Overlay", () => {
       });
     });
   });
+
+  describe("WHEN controlled", () => {
```
Comment (lukeac123) [inline packages/lab/src/__tests__/__e2e__/overlay/Overlay.cy.tsx:119]: tests make sure onOpenChange is called correctly when overlay is in controlled state

Comment (joshwooding) [reply packages/lab/src/__tests__/__e2e__/overlay/Overlay.cy.tsx:119]: This should also test it's called in an uncontrolled state

File: packages/lab/src/__tests__/__e2e__/overlay/Overlay.cy.tsx [hunk 10]
```diff
@@ -112,4 +115,26 @@ describe("GIVEN an Overlay", () => {
...
+    it("THEN it should call onOpenChange on trigger press", () => {
```
Comment (joshwooding) [inline packages/lab/src/__tests__/__e2e__/overlay/Overlay.cy.tsx:120]: NIT: the trigger isn't being pressed.

File: .changeset/plenty-schools-fix.md (hunk 2 above)
Comment (joshwooding) [reply .changeset/plenty-schools-fix.md:5]: We can remove this:

```suggestion
Removed `onClose` prop from `Overlay`. `onOpenChange` is called for events that open/close the overlay.
```

File: packages/lab/stories/overlay/overlay.stories.tsx [hunk 11]
```diff
@@ -55,24 +55,19 @@ const OverlayTemplate = (props: OverlayProps) => {
   );
 };
 
-export const Default = (props: OverlayProps) => {
-  return OverlayTemplate({ id: "overlay-default", ...props });
-};
-
-export const Bottom = (props: OverlayProps) => {
-  return OverlayTemplate({
+export const Bottom = () => {
```
Comment (joshwooding) [inline packages/lab/stories/overlay/overlay.stories.tsx:58]: Is there a reason why we aren't using the standard template implementation e.g. https://storybook.js.org/docs/7.6/api/csf#upgrading-from-csf-2-to-csf-3

File: site/src/examples/overlay/WithActions.tsx [hunk 12]
```diff
@@ -105,24 +105,24 @@ const WithActionsContent = ({ id, onClose }: WithActionsContentProps) => {
...
     <Overlay
-      open={show}
-      onClose={() => {
-        setShow(false);
-      }}
+      open={open}
+      onOpenChange={onOpenChange}
       onKeyDown={(event) => {
-        event.key === "Escape" && setShow(false);
+        event.key === "Escape" && setOpen(false);
```
Comment (joshwooding) [inline site/src/examples/overlay/WithActions.tsx:118]: What does this code do?

File: packages/lab/src/__tests__/__e2e__/overlay/Overlay.cy.tsx [hunk 13]
```diff
@@ -112,4 +118,26 @@ describe("GIVEN an Overlay", () => {
...
+      cy.mount(<Default open={false} onOpenChange={onOpenChangeSpy} />);
+
+      cy.realPress("Tab");
+      cy.realPress("Enter");
+      cy.get("@onOpenChangeSpy").should("have.callCount", 1);
+    });
+
+    it("THEN it should call the onOpenChange on CloseButton click and outside overlay click", () => {
+      const onOpenChangeSpy = cy.stub().as("onOpenChangeSpy");
+      cy.mount(<Default onOpenChange={onOpenChangeSpy} open />);
```
Comment (origami-z) [inline packages/lab/src/__tests__/__e2e__/overlay/Overlay.cy.tsx:134]: is `open` actually a controlled prop? if so, are we actually testing flipping `open` would show/hide the component?

Comment (lukeac123) [reply packages/lab/src/__tests__/__e2e__/overlay/Overlay.cy.tsx]: update the controlled story - thanks :)

File: packages/lab/src/__tests__/__e2e__/overlay/Overlay.cy.tsx [hunk 14]
```diff
@@ -32,26 +32,28 @@ describe("GIVEN an Overlay", () => {
...
+      cy.get("@onOpenChangeSpy").should("have.callCount", 1);
       cy.findByRole("button", { name: /Close Overlay/i }).realClick();
-      cy.get("@closeSpy").should("have.callCount", 1);
       cy.findByRole("dialog").should("not.exist");
+      cy.get("@onOpenChangeSpy").should("have.callCount", 2);
 
       cy.findByRole("button", { name: /Show Overlay/i }).realClick();
       cy.findByRole("dialog").should("be.visible");
-      // dbl click to trigger outside of Overlay click, see SOF post
-      // https://stackoverflow.com/questions/51254946/cypress-does-not-always-executes-click-on-element/58302240#58302240
```
Comment (lukeac123) [inline packages/lab/src/__tests__/__e2e__/overlay/Overlay.cy.tsx:47]: looks like this behaviour has been fixed - https://github.com/cypress-io/cypress/pull/3030

File: packages/lab/src/__tests__/__e2e__/overlay/Overlay.cy.tsx (hunk 9 above)
Comment (lukeac123) [reply packages/lab/src/__tests__/__e2e__/overlay/Overlay.cy.tsx:122]: added uncontrolled testing to test above L34

**Review by Fercas123 — APPROVED** [review]

File: packages/lab/src/overlay/Overlay.tsx (hunk 1 above)
Comment (libertymayc) [reply packages/lab/src/overlay/Overlay.tsx:24]: (reopened this to check when reviewing) I see the event was removed, do we definitely want that re matching other components?

Comment (joshwooding) [reply packages/lab/src/overlay/Overlay.tsx:24]: That change does match the other components... `onOpenChange` everywhere else doesn't take an event.

File: .changeset/plenty-schools-fix.md [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+  const [open, setOpen] = useState(false);
+  const onOpenChange = (newOpen: boolean) => setOpen(newOpen);
+
+  return (
+    <Overlay open={open} onOpenChange={onOpenChange}>
+      <OverlayTrigger>
+        <Button
+          onClick={() => {
+            setShow(true);
+          }}
```
Comment (joshwooding) [inline .changeset/plenty-schools-fix.md:18]:
```suggestion
```

**Review by joshwooding — CHANGES_REQUESTED** [review]
Separate to the PR Overlay.css needs to be aligned to how we're using base components elsewhere. They shouldn't get their own prefix.

File: packages/lab/src/overlay/Overlay.tsx (hunk 1 above)
Comment (libertymayc) [reply packages/lab/src/overlay/Overlay.tsx:24]: ahhh I was looking at an old commit haha ignore :)

File: packages/lab/src/overlay/OverlayPanel.tsx [hunk 16]
```diff
@@ -9,7 +9,7 @@ import { clsx } from "clsx";
 import { useOverlayContext } from "./OverlayContext";
 import { OverlayPanelBase } from "./OverlayPanelBase";
 
-const withBaseName = makePrefixer("saltOverlayPanel");
+const withBaseName = makePrefixer("saltOverlay");
```
Comment (joshwooding) [inline packages/lab/src/overlay/OverlayPanel.tsx:12]:
```suggestion
const withBaseName = makePrefixer("saltOverlayPanel");
```

This should still probably be this

File: packages/lab/stories/overlay/overlay.stories.tsx [hunk 17]
```diff
@@ -211,26 +207,21 @@ const WithActionsContent = ({
...
-      onKeyDown={(event) => {
-        event.key === "Escape" && setShow(false);
-      }}
-      placement="bottom"
-      id={id}
-    >
+    <Overlay open={open} onOpenChange={onChange} placement="bottom" id={id}>
       <OverlayTrigger>
         <Button
           onClick={() => {
```
Comment (joshwooding) [inline packages/lab/stories/overlay/overlay.stories.tsx:223]: Is this needed?

Comment (lukeac123) [reply packages/lab/stories/overlay/overlay.stories.tsx:223]: nope

**Review by Fercas123 — DISMISSED** [review]

File: .changeset/plenty-schools-fix.md [hunk 18]
```diff
@@ -0,0 +1,21 @@
+---
+"@salt-ds/lab": minor
+---
+
+Removed `onClose` prop from `Overlay`, onOpenChange is called for events that open/close the overlay.
```
Comment (joshwooding) [inline .changeset/plenty-schools-fix.md:5]:
```suggestion
Removed the `onClose` prop from `Overlay`, `onOpenChange` is now called for events that open/close the overlay.
```

File: packages/lab/src/overlay/OverlayPanel.tsx (hunk 16 above)
Comment (joshwooding) [reply packages/lab/src/overlay/OverlayPanel.tsx:12]: This wasn't resolved?

File: .changeset/plenty-schools-fix.md (hunk 18 above)
Comment (lukeac123) [reply .changeset/plenty-schools-fix.md:5]: Will fix in a follow up pr

File: packages/lab/src/overlay/Overlay.tsx [hunk 19]
```diff
@@ -91,34 +91,27 @@ export const Overlay = forwardRef<HTMLDivElement, OverlayProps>(
...
           placement,
           context,
           arrowProps,
           floating,
           reference,
-          handleCloseButton,
+          handleCloseButtonClick,
           getFloatingProps,
           getReferenceProps,
         }}
```
Comment (joshwooding) [inline packages/lab/src/overlay/Overlay.tsx:117]: Is the div on line 119 actually needed

File: packages/lab/stories/overlay/overlay.qa.stories.tsx [hunk 20]
```diff
@@ -0,0 +1,53 @@
...
+
+const OverlayTemplate = ({ id, placement }: OverlayProps) => {
+  return (
+    <Overlay id={id} placement={placement} open>
+      <OverlayTrigger>
+        <Button>Show Overlay</Button>
+      </OverlayTrigger>
+      <OverlayPanel>
+        <>
+          <h3 id={`${id}-header`} className="content-heading">
```
Comment (joshwooding) [inline packages/lab/stories/overlay/overlay.qa.stories.tsx:26]: I'm pretty sure these IDs can be removed?

**Review by joshwooding — APPROVED** [review]
Changes to be addressed in another PR

**Review by Fercas123 — APPROVED** [review]

**Pull request merged**